### ISSUE TITLE
eventlet.wsgi.Server.process_request invocation changed

### ISSUE THREAD

**Issue by mattbennett** [eventlet/eventlet#420]
https://github.com/eventlet/eventlet/commit/7f53465578543156e7251e243c0636e087a8445f changes the invocation of https://github.com/eventlet/eventlet/blob/v0.21.0/eventlet/wsgi.py#L740.

The expected contents of the parameter tuple changes from `(<sock>, <address>)` to `(<address>, <sock>, <socket_state>)`. Was this changed for a significant reason? As far as I can make out the socket state is not used and always `STATE_IDLE` after the `accept()`.

I noticed because we run the [Nameko](https://github.com/nameko/nameko) test suite against eventlet master. It would be unfortunate if we had to add a code branch for newer and older versions of Eventlet.

**Comment by temoto** [comment]
> Was this changed for a significant reason?

Yes, see https://github.com/eventlet/eventlet/commit/7f53465578543156e7251e243c0636e087a8445f commit description and https://github.com/eventlet/eventlet/issues/188 for more details.

Was it necessary to change function signature to track connection state?

I'm sorry, it was the best I came up with. Second best option was to rewrite whole thing without `HttpProtocol` class nonsense. Sadly, eventlet wsgi is based on ugly overcomplicated using-classes-in-worst-way stdlib `HTTPServer`/`BaseHTTPRequestHandler`. Sorry for emotional overflow, but read the code, it's counter productive at every second step.

Your best option is to treat `Server.process_request()` as internal and not rely on its existence lest arguments. I can "hide" `process_request` behind underscore if that helps. The way I see WSGI server it only talks `send/recv/close`.

Please continue to test against master, it's super helpful.

> As far as I can make out the socket state is not used...

```
eventlet/wsgi.py:
   32  STATE_IDLE = 'idle'
   33  STATE_REQUEST = 'request'
   34: STATE_CLOSE = 'close'
   35  
   36  __all__ = ['server', 'format_date_time']
   ..
  367          while True:
  368              self.handle_one_request()
  369:             if self.conn_state[2] == STATE_CLOSE:
  370                  self.close_connection = 1
  371              if self.close_connection:
  ...
  931      def _clean_connection(_, conn):
  932          connections.pop(conn[0], None)
  933:         conn[2] = STATE_CLOSE
  934          greenio.shutdown_safe(conn[1])
  935          conn[1].close()
  ...
  954          for cs in six.itervalues(connections):
  955              prev_state = cs[2]
  956:             cs[2] = STATE_CLOSE
  957              if prev_state == STATE_IDLE:
  958                  greenio.shutdown_safe(cs[1])
```

As always, constructive criticism, ideas for improvement are warmly welcome.

**Comment by mattbennett** [comment]
Thanks for the reply. Totally agree re: class structure. I really struggled to find my way around bouncing between all the levels of inheritance. I see where it's mutated now, thank you.

> The way I see WSGI server it only talks `send/recv/close`.

I'm confused about this. What is proper the public interface for the Eventlet WSGI server? Is it possible to use public-only methods and still do the `socket.accept()` outside the API? I guess not, and unfortunately I think Nameko has that requirement

**Comment by temoto** [comment]
@mattbennett we can definitely refactor replaceable `socket.accept`, no big deal.

For now your best options:

* patch `socket.accept` similar to `mock.patch(side_effect=fun)`
* copy-paste `eventlet.wsgi.server`, as you said before, it would require eventlet version branch

**Comment by mattbennett** [comment]
Thank you. Can you give me an example of how you'd patch `socket.accept`?

**Comment by temoto** [comment]
@mattbennett this is how I'd do it

```
def normal_code():
  sock = eventlet.green.socket.listen(...)

  original_accept = sock.accept
  @functools.wraps(sock.accept)
  def patched(self):
    conn = original_accept()
    # what you need
    return conn
  sock.accept = patched

  srv = eventlet.wsgi.Server(sock=sock, ...)
```
